pkg/gardenlet/controller/shootstate/secret/add.go

```diff
@@ -0,0 +1,86 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shootsecret
+
+import (
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "shootsecret"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster, seedCluster cluster.Cluster) error {
+	if r.GardenClient == nil {
+		r.GardenClient = gardenCluster.GetClient()
+	}
+	if r.SeedClient == nil {
+		r.SeedClient = seedCluster.GetClient()
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&corev1.Secret{}, builder.WithPredicates(r.SecretPredicate())).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: *r.Config.ConcurrentSyncs,
+			RecoverPanic:            true,
+			RateLimiter:             workqueue.DefaultControllerRateLimiter(),
+		}).
+		Complete(r)
+}
+
+// SecretPredicate returns the predicates for the secret watch.
+func (r *Reconciler) SecretPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			secret, ok := e.Object.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			return LabelsPredicate(secret.Labels)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			secret, ok := e.Object.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			return LabelsPredicate(secret.Labels)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newSecret, ok := e.ObjectNew.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			return LabelsPredicate(newSecret.Labels)
+		},
+	}
+}
+
+// LabelsPredicate is a function which returns true when the provided labels map suggests that the object is managed by
+// the secrets manager and should be persisted.
+func LabelsPredicate(labels map[string]string) bool {
+	return labels[secretsmanager.LabelKeyManagedBy] == secretsmanager.LabelValueSecretsManager &&
+		labels[secretsmanager.LabelKeyPersist] == secretsmanager.LabelValueTrue
+}
```

Comment:
You can do this much shorter by using this approach: https://github.com/gardener/gardener/blob/0aedc71e4c6ec1e7438ad671d684caca33049b82/pkg/gardenlet/controller/controllerinstallation/care/add.go#L100-L106.

Maybe you can also simplify https://github.com/gardener/gardener/blob/921370a3bf2484ec57fba375a7201003438cc9da/pkg/controllermanager/controller/project/activity/add.go#L117-L154 now that you are on it.

Reply:
https://github.com/gardener/gardener/blob/5d4cd620c96de4b1b7055a706260091585c4df5a/vendor/sigs.k8s.io/controller-runtime/pkg/predicate/predicate.go#L109-L111
only act on the newObject. For project activity we need to consider the label on the oldObject as well.

Reply:
Wait, I just noticed something, Wrote here, https://github.com/gardener/gardener/pull/6667/commits/b47506ae3e7b8040f81236e04bb1c24c4ed2895a#r993234799

Reply:
> https://github.com/gardener/gardener/blob/5d4cd620c96de4b1b7055a706260091585c4df5a/vendor/sigs.k8s.io/controller-runtime/pkg/predicate/predicate.go#L109-L111
> 
> 
> only act on the newObject. For project activity we need to consider the label on the oldObject as well.

The secret reconciled by this controller has nothing to do with the project. It reconciles secrets in seed in shoot control plane name space. So for these secrets to exist in the first place, there needs to be a shoot and if a shoot is there project will not be marked stale.
Am I missing something here?

Reply:
I have not adapted the project controller as suggested above due to the open question from @shafeeqes. 
@rfranzke can you please check what behavior is right here https://github.com/gardener/gardener/pull/6667#discussion_r993234799?

Reply:
Yeah, that's fine @acumino, let's keep this out of this PR and fix separately.